rclpy/rclpy/action/server.py

```diff
@@ -0,0 +1,597 @@
+class ServerGoalHandle:
+    """Goal handle for working with Action Servers."""
+
+    def __init__(self, action_server, goal_info, goal_request):
+        """
+        Accept a new goal with the given action server.
+
+        Instances of this class should only be constructed in the ActionServer class.
+        Instances for accepted goals will be passed to the user-defined goal execution functions.
+
+        If the goal fails to be accepted, then a RuntimeError is raised.
+
+        :param action_server: The ActionServer to accept the goal.
+        :param goal_info: GoalInfo message.
+        :param goal_request: The original goal request message from an ActionClient.
+        """
+        self._handle = _rclpy_action.rclpy_action_accept_new_goal(
+            action_server._handle, goal_info)
+        self._action_server = action_server
+        self._goal_info = goal_info
+        self._goal_request = goal_request
+        self._cancel_requested = False
+        self._result_future = Future()
+        action_server.add_future(self._result_future)
+        self._lock = threading.Lock()
+
+    def __eq__(self, other):
+        return self.goal_id == other.goal_id
+
+    def __ne__(self, other):
+        return self.goal_id != other.goal_id
+
+    @property
+    def request(self):
+        return self._goal_request
+
+    @property
+    def goal_id(self):
+        return self._goal_info.goal_id
+
+    @property
+    def is_active(self):
+        with self._lock:
+            if self._handle is None:
+                return False
+            return _rclpy_action.rclpy_action_goal_handle_is_active(self._handle)
+
+    @property
+    def is_cancel_requested(self):
+        return GoalStatus.STATUS_CANCELING == self.status
+
+    @property
+    def status(self):
+        with self._lock:
+            if self._handle is None:
+                return GoalStatus.STATUS_UNKNOWN
+            return _rclpy_action.rclpy_action_goal_handle_get_status(self._handle)
+
+    def _update_state(self, event):
+        with self._lock:
+            # Ignore updates for already destructed goal handles
+            if self._handle is None:
+                return
+
+            # Update state
+            _rclpy_action.rclpy_action_update_goal_state(self._handle, event.value)
+
+            # Publish state change
+            _rclpy_action.rclpy_action_publish_status(self._action_server._handle)
+
+            # If it's a terminal state, then also notify the action server
+            if not _rclpy_action.rclpy_action_goal_handle_is_active(self._handle):
+                self._action_server.notify_goal_done()
+
+    def execute(self, execute_callback=None):
+        # It's possible that there has been a request to cancel the goal prior to executing.
+        # In this case we want to avoid the illegal state transition to EXECUTING
+        # but still call the users execute callback to let them handle canceling the goal.
+        if not self.is_cancel_requested:
```

Comment:
Does `is_cancel_requested` mean the cancellation was accepted? I think the transition is allowed then. If this only means it was requested then what happens if: a goal to be requested to be canceled, the execute callback is called but the state transition to `EXECUTE` is skipped, then the cancellation request rejected, and finally the execute callback tries to set a terminal state? It looks like it might try to transition from `ACCEPTED -> SUCCEEDED/ABORTED` which isn't an allowed transition.

Reply:
> Does `is_cancel_requested` mean the cancellation was accepted?

Yes. The cancellation request was accepted and the goal is not in the `CANCELING` state.

https://github.com/ros2/rclpy/blob/b03718678f5a60948175f7e7da8b9a6e7a6e95c3/rclpy/rclpy/action/server.py#L102-L103